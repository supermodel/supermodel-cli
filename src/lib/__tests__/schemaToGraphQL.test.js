const path = require('path')
const fs = require('fs')
const schemaToGraphQL = require('../schemaToGraphQL')

function matchSchema(schema) {
  expect(schemaToGraphQL(schema)).toMatchSnapshot();
}

describe('schemaToGraphQL', () => {
  const CarSchema = {
    "$id": "http://supermodel.io/factory/Car",
    "title": "Article",
    "type": "object",
    "properties": {
      "manufacturer": {
        "type": "string"
      },
      "type": {
        "type": "string"
      },
      "weight": {
        "type": "integer"
      },
      "length": {
        "type": "number"
      },
      "available": {
        "type": "boolean"
      }
    }
  }

  const EngineSchema = {
    "$id": "http://supermodel.io/factory/parts/Engine",
    "title": "Engine",
    "type": "object",
    "properties": {
      "volume": {
        "type": "number"
      },
      "power": {
        "type": "integer"
      }
    }
  }

  const ManufacturerSchema = {
    "$id": "http://supermodel.io/factory/Manufacturer",
    "title": "Manufacturer",
    "type": "object",
    "properties": {
      "name": {
        "type": "string"
      }
    }
  }

  test('converts simple object', () => {
    matchSchema(CarSchema)
  })

  test('converts required property', () => {
    matchSchema({
      ...CarSchema,
      required: [
        'manufacturer',
        'length'
      ]
    })
  })

  test('converts enum', () => {
    matchSchema({
      ...EngineSchema,
      properties: {
        ...EngineSchema.properties,
        fuel: {
          type: "string",
          enum: [
            "gasoline",
            "diesel",
            "lpg",
            "cng"
          ]
        }
      }
    })
  })

  test('converts inline object in property', () => {
    const { $id, ...EngineSchemaWithoutId } = EngineSchema

    matchSchema({
      ...CarSchema,
      properties: {
        ...CarSchema.properties,
        engine: EngineSchemaWithoutId
      }
    })
  })

  describe('resolves refs', () => {
    test('by id', () => {
      matchSchema({
        ...CarSchema,
        properties: {
          ...CarSchema.properties,
          engine: {
            $ref: EngineSchema.$id
          }
        },
        definitions: {
          [EngineSchema.$id]: EngineSchema
        }
      })
    })

    test('by relative path', () => {
      matchSchema({
        ...CarSchema,
        properties: {
          ...CarSchema.properties,
          engine: {
            $ref: 'parts/Engine'
          }
        },
        definitions: {
          [EngineSchema.$id]: EngineSchema
        }
      })
    })

    test('by #/', () => {
      matchSchema({
        ...CarSchema,
        properties: {
          ...CarSchema.properties,
          engine: {
            $ref: '#/definitions/parts/Engine'
          }
        },
        definitions: {
          'parts/Engine': EngineSchema
        }
      })
    })

    test('works with nested ref objects', () => {
      const EngineSchemaWithManufaturer = {
        ...EngineSchema,
        properties: {
          ...EngineSchema.properties,
          manufacturer: {
            $ref: ManufacturerSchema.$id
          }
        }
      }

      matchSchema({
        ...CarSchema,
        properties: {
          ...CarSchema.properties,
          manufacturer: {
            $ref: ManufacturerSchema.$id
          },
          engine: {
            $ref: EngineSchemaWithManufaturer.$id
          }
        },
        definitions: {
          [EngineSchemaWithManufaturer.$id]: EngineSchemaWithManufaturer,
          [ManufacturerSchema.$id]: ManufacturerSchema
        }
      })
    })
  })

  test('complex schema', () => {
    const content = fs.readFileSync(path.resolve(__dirname, './__fixtures__/Article.json'))
    const schema = JSON.parse(content.toString())
    matchSchema(schema)
  })
})
