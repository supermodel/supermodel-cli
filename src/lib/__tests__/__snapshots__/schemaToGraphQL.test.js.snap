// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schemaToGraphQL complex schema 1`] = `
"type DavidAdidasProductArticle {
  articleNumber: String
  modelNumber: String
  articleDescription: String
  gender: DavidAdidasProductUtilsCodeDescription
  brand: DavidAdidasProductUtilsCodeDescription
  productType: DavidAdidasProductUtilsCodeDescription
  ageGroup: DavidAdidasProductUtilsCodeDescription
  colorway: DavidAdidasProductArticleColorway
  colors: [DavidAdidasProductColor]
  seasons: [DavidAdidasProductArticleSeason]
  skus: [DavidAdidasProduct]
  assets: [DavidAdidasProductAsset]
  catalog: DavidAdidasProductCatalog
  related: [DavidAdidasProductArticle]
}

type DavidAdidasProductUtilsCodeDescription {
  code: String
  description: String
}

type DavidAdidasProductArticleColorway {
  name: String
  description: String
}

type DavidAdidasProductColor {
  color1: DavidAdidasProductUtilsCodeDescription
  color2: DavidAdidasProductUtilsCodeDescription
  color3: DavidAdidasProductUtilsCodeDescription
  color4: DavidAdidasProductUtilsCodeDescription
}

type DavidAdidasProductArticleSeason {
  seasonName: String
  carryOver: Boolean
  retailIntroDate: String
  retailExitDate: String
}

type DavidAdidasProduct {
  EAN: String
  UPC: String
  srcSizeCD3: String
  srcSizeScale: String
  srcSizeDescription: String
  conversionIDs: [DavidAdidasProductConversionDs]
}

type DavidAdidasProductAsset {
  assetType: DavidAdidasProductAssetAssetTypeEnum!
  viewGroup: String!
  view: String!
  assetCategory: String!
  usageTerms: [String]!
  assetHref: String!
  image: DavidAdidasProductAssetImage
  video: DavidAdidasProductAssetVideo
  model: DavidAdidasProductAssetModel
}

type DavidAdidasProductCatalog {
  catalogueCopy: DavidAdidasProductCatalogCatalogCopy
  featureAndBenefits: [DavidAdidasProductCatalogFeatureAndBenefit]
}

type DavidAdidasProductConversionDs {
  localSize: String
  saleSizeConvID: String
  saleSizeScale: String
  saleSizeAbbr: String
  saleSizeDesc: String
}

enum DavidAdidasProductAssetAssetTypeEnum {
  image
  video
  other
}

type DavidAdidasProductAssetImage {
  style: String
  width: Int
  height: Int
  bgColor: String
  representation: String
}

type DavidAdidasProductAssetVideo {
  fileSize: Int
}

type DavidAdidasProductAssetModel {
  waist: String
  chest: String
  height: String
  productSize: String
}

type DavidAdidasProductCatalogCatalogCopy {
  acsModelName: String
  b2cCopy: String
  b2cCopyShort: String
  b2bCopy: String
  b2bCopyShort: String
  materialComposition: String
  materialConstruction: String
  bullets1: String
  bullets2: String
  bullets3: String
  bullets4: String
  bullets5: String
  bullets6: String
  bulletMaterial: String
}

type DavidAdidasProductCatalogFeatureAndBenefit {
  featureAndBenefitCode: String
  businessCode: String
  shoePart: String
  featureName: String
  featureDescription: String
  logoName: String
  logoHref: String
}"
`;

exports[`schemaToGraphQL converts enum 1`] = `
"type SupermodelFactoryPartsEngine {
  volume: Float
  power: Int
  fuel: SupermodelFactoryPartsEngineFuelEnum
}

enum SupermodelFactoryPartsEngineFuelEnum {
  gasoline
  diesel
  lpg
  cng
}"
`;

exports[`schemaToGraphQL converts inline object in property 1`] = `
"type SupermodelFactoryCar {
  manufacturer: String
  type: String
  weight: Int
  length: Float
  available: Boolean
  engine: SupermodelFactoryCarEngine
}

type SupermodelFactoryCarEngine {
  volume: Float
  power: Int
}"
`;

exports[`schemaToGraphQL converts required property 1`] = `
"type SupermodelFactoryCar {
  manufacturer: String!
  type: String
  weight: Int
  length: Float!
  available: Boolean
}"
`;

exports[`schemaToGraphQL converts simple object 1`] = `
"type SupermodelFactoryCar {
  manufacturer: String
  type: String
  weight: Int
  length: Float
  available: Boolean
}"
`;

exports[`schemaToGraphQL resolves refs by #/ 1`] = `
"type SupermodelFactoryCar {
  manufacturer: String
  type: String
  weight: Int
  length: Float
  available: Boolean
  engine: SupermodelFactoryPartsEngine
}

type SupermodelFactoryPartsEngine {
  volume: Float
  power: Int
}"
`;

exports[`schemaToGraphQL resolves refs by id 1`] = `
"type SupermodelFactoryCar {
  manufacturer: String
  type: String
  weight: Int
  length: Float
  available: Boolean
  engine: SupermodelFactoryPartsEngine
}

type SupermodelFactoryPartsEngine {
  volume: Float
  power: Int
}"
`;

exports[`schemaToGraphQL resolves refs by relative path 1`] = `
"type SupermodelFactoryCar {
  manufacturer: String
  type: String
  weight: Int
  length: Float
  available: Boolean
  engine: SupermodelFactoryPartsEngine
}

type SupermodelFactoryPartsEngine {
  volume: Float
  power: Int
}"
`;

exports[`schemaToGraphQL resolves refs works with nested ref objects 1`] = `
"type SupermodelFactoryCar {
  manufacturer: SupermodelFactoryManufacturer
  type: String
  weight: Int
  length: Float
  available: Boolean
  engine: SupermodelFactoryPartsEngine
}

type SupermodelFactoryManufacturer {
  name: String
}

type SupermodelFactoryPartsEngine {
  volume: Float
  power: Int
  manufacturer: SupermodelFactoryManufacturer
}"
`;
